src/key_info_managers/mod.rs

```diff
@@ -224,4 +229,121 @@ impl dyn ManageKeyInfo + Send + Sync {
             Ok(())
         }
     }
+
+    /// Remove the key represented by a key triple and return the stored info.
+    ///
+    /// # Errors
+    ///
+    /// If the key does not exist, PsaErrorDoesNotExist is returned. If any other error occurs,
+    /// KeyInfoManagerError is returned.
+    pub fn remove_key(
+        &mut self,
+        key_triple: &KeyTriple,
+    ) -> parsec_interface::requests::Result<KeyInfo> {
+        match self.remove(key_triple) {
+            Ok(Some(key_info)) => Ok(key_info),
+            Ok(None) => Err(ResponseStatus::PsaErrorDoesNotExist),
+            Err(string) => Err(to_response_status(string)),
+        }
+    }
+
+    /// Insert key info for a given triple.
+    ///
+    /// # Errors
+    ///
+    /// If the key triple already existed in the KIM, PsaErrorAlreadyExists is returned. For
+    /// any other error occurring in the KIM, KeyInfoManagerError is returned.
+    pub fn insert_key_info<T: Serialize>(
```

Comment:
I don't want to sound negative, but just discussing the possibility of *not* having to add a new structure and add more methods on the same type. I think it might be possible to do:
```rust
    pub fn insert_key_info<T: Serialize>(
        kim: &Arc<RwLock<Self>>,
        key_triple: KeyTriple,
        key_id: &T,
        attributes: Attributes,
    ) -> parsec_interface::requests::Result<()> {
        let mut key_info_manager = kim
            .write()
            .expect("Key Info Manager lock poisoned");
        let key_info = KeyInfo {
            id: bincode::serialize(key_id)?,
            attributes,
        };

        match key_info_manager.insert(key_triple, key_info) {
            Ok(None) => Ok(()),
            Ok(Some(_)) => Err(ResponseStatus::PsaErrorAlreadyExists),
            Err(string) => Err(to_response_status(string)),
        }
    }
```
So that providers call it via `insert_key_info(self.key_info_store, ...)` without having to deal with the new structure?

Reply:
Could do, but at the moment we have to handle the monstruosity which is `Arc<RWLock<dyn ..........>>` which is one of the things that I'm trying to remove - it's not necessary for providers to know or care about all that, yet we have to do it everywhere. Sure, we could just hide part of it using some methods, but in the grand scheme of things it looks like some complexity that can be hidden in a new structure.

Reply:
I see what you mean, although if you have methods taking this `Arc<RwLock<...>>` type, then providers should not have to care about it and can call via `insert_key_info(self.key_info_store, ...)` and others.
I can see the complexity of understand the type and its methods when writing a new provider, but there is also the complexity of having one more type to maintain with methods that are very similar to those on the impl?
Another advantage of using the method I described is that maybe you can only keep one single version of those methods, the one having `Arc<RwLock<...>>` as parameter

Reply:
> but there is also the complexity of having one more type to maintain with methods that are very similar to those on the impl

I can just remove the methods on the `impl` 😛 there is no reason to have them there as well if we have the client. Another reason why I'd prefer the client is because we don't have to deal with putting the correct `ProviderID` when dealing with the KIM

Reply:
> I can just remove the methods on the impl

Yes I would agree with that.
I think I am ok with it. Could we somehow make what you call the "client" **the** KeyInfoManager? As in having the idea that the `ManageKeyInfo` and its implementations are private but the "client" or manager is the only thing that is public to provider?

Reply:
> Could we somehow make what you call the "client" the KeyInfoManager?

Yes, I can try, but it'll be a big change :D Also, I need things to be public at least within the crate, because the service builder still needs to instantiate things. But for Providers it makes sense to never see anything but the "client"

> is it possible to return a T with a get_key_info method?

I'll try, I keep forgetting what the rules for generics are, whether returning multiple types is allowed (but I seem to remember it was)

Reply:
> the service builder still needs to instantiate things

Maybe then the service builder could build that new structure instead of what we built before and pass it to the providers directly? Inside the builder of that structure we could build the implementations, that way we would hide everything!

Agree it would be a big change 🧑‍🔧 But happy to review a big PR

Reply:
There's a problem with this approach - the service builder can't share the `Arc<...>` between instantiations of `KeyInfoManager`. I have some ideas, though